File: PreStage-Mac/Sources/PreStage/Support/PreviewRenderGeometry.swift
import CoreGraphics
import Foundation

enum PreviewRenderGeometry {
    static func imageRect(aspectRatio: CGFloat, in size: CGSize, scale: CGFloat) -> CGRect {
        guard aspectRatio > 0, size.width > 0, size.height > 0 else {
            return CGRect(origin: .zero, size: size)
        }

        let containerAspect = size.width / size.height
        let fittedSize: CGSize
        if aspectRatio > containerAspect {
            fittedSize = CGSize(width: size.width, height: size.width / aspectRatio)
        } else {
            fittedSize = CGSize(width: size.height * aspectRatio, height: size.height)
        }

        return pixelAligned(
            CGRect(
                x: (size.width - fittedSize.width) / 2,
                y: (size.height - fittedSize.height) / 2,
                width: fittedSize.width,
                height: fittedSize.height
            ),
            scale: scale
        )
    }

    static func cropRect(in imageRect: CGRect, aspectRatio: CGFloat, scale: CGFloat) -> CGRect {
        guard aspectRatio > 0, imageRect.width > 0, imageRect.height > 0 else {
            return .zero
        }

        let imageAspect = imageRect.width / imageRect.height
        let cropSize: CGSize
        if aspectRatio > imageAspect {
            cropSize = CGSize(width: imageRect.width, height: imageRect.width / aspectRatio)
        } else {
            cropSize = CGSize(width: imageRect.height * aspectRatio, height: imageRect.height)
        }

        return pixelAligned(
            CGRect(
                x: imageRect.midX - cropSize.width / 2,
                y: imageRect.midY - cropSize.height / 2,
                width: cropSize.width,
                height: cropSize.height
            ),
            scale: scale
        )
    }

    static func maskRects(imageRect: CGRect, cropRect: CGRect, scale: CGFloat) -> [CGRect] {
        let image = pixelAligned(imageRect, scale: scale)
        let crop = pixelAligned(cropRect, scale: scale)
        guard image.width > 0, image.height > 0, crop.width > 0, crop.height > 0 else {
            return []
        }

        let bleed = 1 / max(scale, 1)
        let top = CGRect(
            x: image.minX - bleed,
            y: image.minY - bleed,
            width: image.width + bleed * 2,
            height: max(0, crop.minY - image.minY) + bleed * 2
        )
        let bottom = CGRect(
            x: image.minX - bleed,
            y: crop.maxY - bleed,
            width: image.width + bleed * 2,
            height: max(0, image.maxY - crop.maxY) + bleed * 2
        )
        let left = CGRect(
            x: image.minX - bleed,
            y: crop.minY - bleed,
            width: max(0, crop.minX - image.minX) + bleed * 2,
            height: crop.height + bleed * 2
        )
        let right = CGRect(
            x: crop.maxX - bleed,
            y: crop.minY - bleed,
            width: max(0, image.maxX - crop.maxX) + bleed * 2,
            height: crop.height + bleed * 2
        )

        return [top, bottom, left, right]
            .filter { $0.width > bleed && $0.height > bleed }
            .map { pixelAligned($0, scale: scale) }
    }

    static func pixelAligned(_ rect: CGRect, scale: CGFloat) -> CGRect {
        let scale = max(scale, 1)
        let minX = floor(rect.minX * scale) / scale
        let minY = floor(rect.minY * scale) / scale
        let maxX = ceil(rect.maxX * scale) / scale
        let maxY = ceil(rect.maxY * scale) / scale
        return CGRect(x: minX, y: minY, width: max(0, maxX - minX), height: max(0, maxY - minY))
    }
}

enum PreviewSourceGeometry {
    static func aspectRatio(for item: MediaItem, previewURL: URL) -> CGFloat {
        if previewURL != item.url, let previewAspect = imageAspectRatio(at: previewURL) {
            return previewAspect
        }
        if let itemAspect = item.displayAspectRatio {
            return itemAspect
        }
        if let previewAspect = imageAspectRatio(at: previewURL) {
            return previewAspect
        }
        return 1
    }

    static func imageAspectRatio(at url: URL) -> CGFloat? {
        PreviewDecodeService.shared.imageAspectRatio(at: url)
    }

    static func imagePixelSize(at url: URL) -> CGSize? {
        PreviewDecodeService.shared.imagePixelSize(at: url)
    }

    static func supportsDirectRasterPreview(url: URL) -> Bool {
        PreviewDecodeService.shared.supportsDirectRasterPreview(url: url)
    }
}

File: PreStage.Core/Imaging/PreviewRenderGeometry.cs
namespace PreStage.Core.Imaging;

public readonly struct GeometryRect
{
    public double X { get; }
    public double Y { get; }
    public double Width { get; }
    public double Height { get; }

    public double MinX => X;
    public double MinY => Y;
    public double MaxX => X + Width;
    public double MaxY => Y + Height;
    public double MidX => X + Width / 2.0;
    public double MidY => Y + Height / 2.0;

    public GeometryRect(double x, double y, double width, double height)
    {
        X = x;
        Y = y;
        Width = Math.Max(0, width);
        Height = Math.Max(0, height);
    }

    public static GeometryRect FromMinMax(double minX, double minY, double maxX, double maxY)
    {
        return new GeometryRect(minX, minY, maxX - minX, maxY - minY);
    }
}

public readonly struct GeometrySize
{
    public double Width { get; }
    public double Height { get; }

    public GeometrySize(double width, double height)
    {
        Width = width;
        Height = height;
    }
}

public static class PreviewRenderGeometry
{
    public static GeometryRect ImageRect(double aspectRatio, GeometrySize containerSize, double scale)
    {
        if (aspectRatio <= 0 || containerSize.Width <= 0 || containerSize.Height <= 0)
            return new GeometryRect(0, 0, containerSize.Width, containerSize.Height);

        var containerAspect = containerSize.Width / containerSize.Height;
        double fittedWidth, fittedHeight;

        if (aspectRatio > containerAspect)
        {
            fittedWidth = containerSize.Width;
            fittedHeight = containerSize.Width / aspectRatio;
        }
        else
        {
            fittedWidth = containerSize.Height * aspectRatio;
            fittedHeight = containerSize.Height;
        }

        return PixelAlign(new GeometryRect(
            (containerSize.Width - fittedWidth) / 2.0,
            (containerSize.Height - fittedHeight) / 2.0,
            fittedWidth,
            fittedHeight
        ), scale);
    }

    public static GeometryRect CropRect(GeometryRect imageRect, double aspectRatio, double scale)
    {
        if (aspectRatio <= 0 || imageRect.Width <= 0 || imageRect.Height <= 0)
            return new GeometryRect(0, 0, 0, 0);

        var imageAspect = imageRect.Width / imageRect.Height;
        double cropWidth, cropHeight;

        if (aspectRatio > imageAspect)
        {
            cropWidth = imageRect.Width;
            cropHeight = imageRect.Width / aspectRatio;
        }
        else
        {
            cropWidth = imageRect.Height * aspectRatio;
            cropHeight = imageRect.Height;
        }

        return PixelAlign(new GeometryRect(
            imageRect.MidX - cropWidth / 2.0,
            imageRect.MidY - cropHeight / 2.0,
            cropWidth,
            cropHeight
        ), scale);
    }

    public static GeometryRect[] MaskRects(GeometryRect imageRect, GeometryRect cropRect, double scale)
    {
        var image = PixelAlign(imageRect, scale);
        var crop = PixelAlign(cropRect, scale);

        if (image.Width <= 0 || image.Height <= 0 || crop.Width <= 0 || crop.Height <= 0)
            return [];

        var bleed = 1.0 / Math.Max(scale, 1.0);

        var top = new GeometryRect(
            image.MinX - bleed,
            image.MinY - bleed,
            image.Width + bleed * 2.0,
            Math.Max(0, crop.MinY - image.MinY) + bleed * 2.0
        );

        var bottom = new GeometryRect(
            image.MinX - bleed,
            crop.MaxY - bleed,
            image.Width + bleed * 2.0,
            Math.Max(0, image.MaxY - crop.MaxY) + bleed * 2.0
        );

        var left = new GeometryRect(
            image.MinX - bleed,
            crop.MinY - bleed,
            Math.Max(0, crop.MinX - image.MinX) + bleed * 2.0,
            crop.Height + bleed * 2.0
        );

        var right = new GeometryRect(
            crop.MaxX - bleed,
            crop.MinY - bleed,
            Math.Max(0, image.MaxX - crop.MaxX) + bleed * 2.0,
            crop.Height + bleed * 2.0
        );

        return new[] { top, bottom, left, right }
            .Where(r => r.Width > bleed && r.Height > bleed)
            .Select(r => PixelAlign(r, scale))
            .ToArray();
    }

    public static GeometryRect PixelAlign(GeometryRect rect, double scale)
    {
        scale = Math.Max(scale, 1.0);
        var minX = Math.Floor(rect.MinX * scale) / scale;
        var minY = Math.Floor(rect.MinY * scale) / scale;
        var maxX = Math.Ceiling(rect.MaxX * scale) / scale;
        var maxY = Math.Ceiling(rect.MaxY * scale) / scale;
        return GeometryRect.FromMinMax(minX, minY, maxX, maxY);
    }
}
